src/utils.jl

```diff
@@ -0,0 +1,80 @@
+"""
+Efficiently reduce a BitArray with or (|)
+"""
+function or(B::BitArray)
+    for b in B
+        if b
+            return true
+        end
+    end
+    return false
+end
+
+function or(B::NTuple{N,Bool}) where N
+    for b in B
+        if b
+            return true
+        end
+    end
+    return false
+end
+
+"""
+Efficiently reduce a BitArray with and (&)
+"""
+function and(B::BitArray)
+    for b in B
+        if !b
+            return false
+        end
+    end
+    return true
+end
+
+function and(B::NTuple{N, Bool}) where N
+    for b in B
+        if !b
+            return false
+        end
+    end
+    return true
+end
+
+"""
+A function to generate the correct permutation to flip an array of dimension `N` to be orthogonal to `dim`, when the Nth dimension is a singleton
+"""
```

Comment:
I don't get this

Reply:
it's useful if you have an Array{T,N-1} `A` that you want to concatanate with an Array{T,N} `B`.
You need to give `A` an extra dimension of length 1 so that it's `N` dimensional, and then permute its dimensions such that it is pointed orthogonally to the dimension that you want to join it to `B` along, i.e. the length 1 dimension is in position `dim`.
```
A = rand(2,6,7)
B = rand(2,5,6,7) #we want to join A to be along dimension 2
A_N = add_dims(A, 1)

A_permuted = permutedims(A_N, orth_perm(4, 2))
cat(A_permuted, B, dims=2)
```

Reply:
so it's just reshape?

Reply:
Well, when you put it that way :sweat_smile:...
I hadn't realised that this was equivalent